os/drivers/audio/syu645b.c

```diff
@@ -0,0 +1,1132 @@
+        }
+
+        return ret;
+}
+
+/************************************************************************************
+ * Name: syu645b_exec_i2c_script
+ *
+ * Description:
+ *   Executes given script through i2c to configuure SYU645B device.
+ *
+ ************************************************************************************/
+
+static int syu645b_exec_i2c_script(FAR struct syu645b_dev_s *priv, t_codec_init_script_entry *script, uint32_t size)
+{
+	uint32_t i;
+        uint16_t ret = 0;
+
+         for (i = 0; i < size; i++) {
+
+		if (script[i].type == SYU645B_REG_D_2BYTE) {
+			ret = syu645b_writereg_2byte(priv, script[i].addr, script[i].val[0]);
+                	if (script[i].delay > 0) {
+                		up_mdelay(script[i].delay);
+                	}
+        	} else if (script[i].type == SYU645B_REG_D_3BYTE) {
+			ret = syu645b_writereg_3byte(priv, script[i].addr, script[i].val);
+                	if (script[i].delay > 0) {
+                		up_mdelay(script[i].delay);
+                	}
+		} else if (script[i].type == SYU645B_REG_D_5BYTE) {
+			ret = syu645b_writereg_5byte(priv, script[i].addr, script[i].val);
+                	if (script[i].delay > 0) {
+                		up_mdelay(script[i].delay);
+                	}
+		}
+	}
+
+        return ret;
+}
+
+/************************************************************************************
+ * Name: syu645b_takesem
+ *
+ * Description:
+ *  Take a semaphore count, handling the nasty EINTR return if we are interrupted
+ *  by a signal.
+ *
+ ************************************************************************************/
+static void syu645b_takesem(sem_t *sem)
+{
+        int ret;
+
+        do {
+                ret = sem_wait(sem);
+                DEBUGASSERT(ret == 0 || errno == EINTR);
+        } while (ret < 0);
+}
+
+/************************************************************************************
+ * Name: syu645b_setvolume
+ *
+ * Description:
+ *   Set the right and left volume values in the SYU645B device based on the current
+ *   volume and balance settings.
+ *
+ ************************************************************************************/
+#ifndef CONFIG_AUDIO_EXCLUDE_VOLUME
+static void syu645b_setvolume(FAR struct syu645b_dev_s *priv)
+{
+        /* if no audio device object return */
+        if (!priv) {
+                auddbg("Error, Device's private data Not available\n");
+                return;
+        }
+
+        if (priv->running) {
+                audvdbg("volume=%u mute=%u\n", priv->volume, priv->mute);
+                if (priv->mute) {
+			syu645b_exec_i2c_script(priv, codec_init_mute_on_script, sizeof(codec_init_mute_on_script) / sizeof(t_codec_init_script_entry));
+                } else {
+			if (priv->volume == 0) {
+				codec_set_master_volume_script[0].val[0] = 0;
+			} else {
+				codec_set_master_volume_script[0].val[0] = SYU645B_HW_VOL_MAX - (SYU645B_SPK_VOL_MAX - priv->volume) * SYU645B_HW_VOL_STEP;
```

Comment:
please check volume logic in datasheet more. Media framework has 10 of volume level, volume level should be adjust properly.

Reply:
What I meaning is below.
volume 0 : mute
volume 1 : 10% of max volume
volume 2: 20% of max volume

So we need extra algorithm.